pkg/cmd/roachtest/tests/tpcc.go

```diff
@@ -63,6 +68,107 @@ func rampDuration(isLocal bool) time.Duration {
+) (roachtestutil.AggregatedPerfMetrics, error) {
+	// Group metrics by warehouse number
+	metricsByWarehouse := make(map[int64]*roachtestutil.HistogramMetric)
+	for _, metric := range histogramMetrics.Summaries {
+		// Extract warehouse label value
+		var warehouse string
+		for _, label := range metric.Labels {
+			if label.Name == warehouseLabelString {
+				warehouse = label.Value
+				break
+			}
+		}
+		if warehouse == "" {
+			continue
+		}
+
+		warehouseNum, err := strconv.ParseInt(warehouse, 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed parsing warehouse number %q", warehouse)
+		}
+
+		// Add metric to warehouse group
+		if metrics, exists := metricsByWarehouse[warehouseNum]; exists {
+			metrics.Summaries = append(metrics.Summaries, metric)
+		} else {
+			metricsByWarehouse[warehouseNum] = &roachtestutil.HistogramMetric{
+				Summaries: []*roachtestutil.HistogramSummaryMetric{metric},
+			}
+		}
+	}
+
+	// Calculate TPMC for each warehouse and find max efficient warehouse
+	tpmcByWarehouse := make(map[int64]float64)
+	var maxEfficientWarehouse int64
+
+	for warehouseNum, metrics := range metricsByWarehouse {
+		var totalCount, totalElapsed int64
+		for _, metric := range metrics.Summaries {
+			if metric.Name == newOrderTableName {
+				totalCount += metric.TotalCount
+				totalElapsed += metric.TotalElapsed
+			}
+		}
+
+		// Calculate TPMC (transactions per minute completed)
+		tpmc := (float64(totalCount) * 60000.0) / float64(totalElapsed)
+		tpmcByWarehouse[warehouseNum] = tpmc
+
+		// Check if efficiency meets threshold
+		efficiency := (tpmc / (tpmcPerWarehouseRatio * float64(warehouseNum))) * 100
+		if efficiency > efficiencyThreshold && warehouseNum > maxEfficientWarehouse {
+			maxEfficientWarehouse = warehouseNum
+		}
+	}
+
+	// Build aggregated metrics
+	var aggregatedMetrics roachtestutil.AggregatedPerfMetrics
+	for warehouseNum, tpmc := range tpmcByWarehouse {
+		warehouseLabel := &roachtestutil.Label{
+			Name:  warehouseLabelString,
+			Value: strconv.FormatInt(warehouseNum, 10),
+		}
+
+		aggregatedMetrics = append(aggregatedMetrics,
+			&roachtestutil.AggregatedMetric{
+				Name:             fmt.Sprintf("%s_tpmc", testName),
+				Value:            roachtestutil.MetricPoint(tpmc),
+				Unit:             "txn/min",
+				IsHigherBetter:   true,
+				AdditionalLabels: []*roachtestutil.Label{warehouseLabel},
+			},
+			&roachtestutil.AggregatedMetric{
+				Name:             fmt.Sprintf("%s_efficiency", testName),
+				Value:            roachtestutil.MetricPoint((tpmc / (tpmcPerWarehouseRatio * float64(warehouseNum))) * 100),
+				Unit:             "percentage",
+				IsHigherBetter:   true,
+				AdditionalLabels: []*roachtestutil.Label{warehouseLabel},
+			},
+		)
+	}
+
+	// Add max efficient warehouse metric
+	aggregatedMetrics = append(aggregatedMetrics, &roachtestutil.AggregatedMetric{
+		Name:           fmt.Sprintf("%s_max_efficient_warehouse", testName),
```

Comment:
Nit: `efficient` seems redundant since that's the only report max, and it's always evaluated according to `PassingEfficiency`.